File: dsa/concepts/12_tries/java/L211_DesignAddAndSearchWordsDataStructure.java
import com.algos.topics.models.TrieNode;

public class L211_DesignAddAndSearchWordsDataStructure {

	public static void main(String[] args) {
		L211_DesignAddAndSearchWordsDataStructure obj = new L211_DesignAddAndSearchWordsDataStructure();

		obj.addWord("bad");
		obj.addWord("dad");
		obj.addWord("mad");

		System.out.println(obj.search("pad"));
		System.out.println(obj.search("bad"));
		System.out.println(obj.search(".ad"));
		System.out.println(obj.search("b.."));
	}

    // Problem: https://leetcode.com/problems/design-add-and-search-words-data-structure/
    // Search words with '.' as wildcard
    // Idea: Use Trie to store words
    // For seach if char is '.'. Then loop over all chars in current map and check if any recursion returns true. You need to skip one char from work fow '.'.
	TrieNode root;
    public L211_DesignAddAndSearchWordsDataStructure() {
        root = new TrieNode();
    }

	public void addWord(String word) {
        TrieNode curr = root;
        for(char ch: word.toCharArray()){
            curr.map.putIfAbsent(ch, new TrieNode());
            curr = curr.map.get(ch);
        }
        curr.endOfWord = true;
    }

    public boolean search(String word) {
        return search(word, 0, root);   // when search char is '.' then we need to skip one char from word. Hence we need to pass index as well.
    }

    private boolean search(String word, int ind, TrieNode curr){
        for(int i=ind; i<word.length(); i++){
            char ch = word.charAt(i);
            if(ch != '.'){
                if(!curr.map.containsKey(ch))
                    return false;
                curr = curr.map.get(ch);
            }else{                                      // if char is '.'
                for(TrieNode child: curr.map.values()){
                    if(search(word, i+1, child)){       // i+1 as we need to skip one char from word.
                        return true;
                    }
                }
                return false;
            }
        }
        return curr.endOfWord;
    }

}

File: dsa/concepts/12_tries/python/l211_design_add_and_search_words_data_structure.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.end_of_word = False


class WordDictionary:
    def __init__(self):
        self.root = TrieNode()

    """
    Problem: https://leetcode.com/problems/design-add-and-search-words-data-structure/
    Search words with '.' as wildcard
    Idea: Use Trie to store words
    For search if char is '.', then loop over all chars in current map and check
    if any recursion returns true. You need to skip one char from word for '.'
    """

    def add_word(self, word: str) -> None:
        curr = self.root
        for ch in word:
            curr.children.setdefault(ch, TrieNode())
            curr = curr.children[ch]
        curr.end_of_word = True

    def search(self, word: str) -> bool:
        # when search char is '.' then we need to skip one char from word
        # hence we need to pass index as well
        return self._search(word, 0, self.root)

    def _search(self, word: str, index: int, node: TrieNode) -> bool:
        for i in range(index, len(word)):
            ch = word[i]

            if ch != '.':
                if ch not in node.children:
                    return False
                node = node.children[ch]
            else:
                # if char is '.', try all possible children
                for child in node.children.values():
                    if self._search(word, i + 1, child):  # skip one char
                        return True
                return False

        return node.end_of_word


if __name__ == "__main__":
    wd = WordDictionary()

    wd.add_word("bad")
    wd.add_word("dad")
    wd.add_word("mad")

    print(wd.search("pad"))   # False
    print(wd.search("bad"))   # True
    print(wd.search(".ad"))   # True
    print(wd.search("b.."))   # True
